### PR TITLE
Add nvprims.var_mean

### LINKED ISSUE #83506: Python custom op registration doesn't broadcast optional int arguments
### 🐛 Describe the bug

Try the following code. `dim` argument when passed a single integer should internally be expanded to a list of size 1 with given int. It doesn't work when the argument is specified to be optional and works as expected without `?`.

```py
import torch

test = torch.library.Library("test", "DEF")
test_impl = torch.library.Library("test", "IMPL", "CompositeExplicitAutograd")

schema = "test(int[1]? dim) -> Tensor"

test.define(schema)
test_impl.impl("test", lambda dim=None: torch.empty(dim))

try:
    torch.ops.test.test(dim=2)
except RuntimeError as e:
    print(e)

# But the following works:
import torch

schema = "test_non_optional(int[1] dim) -> Tensor"

test.define(schema)
test_impl.impl("test_non_optional", lambda dim=None: torch.empty(dim))

try:
    torch.ops.test.test_non_optional(dim=2)
except RuntimeError as e:
    print(e)
```

### Versions

Latest master.

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR adds nvfuser-specific primitive - `var_mean`.
Interpretation `torch.var_mean` -> `torch.ops.nvprims.var_mean` is handled by `TorchRefsNvfuserCapabilityMode` context manager.

I moved some helper code from `_prims/__init__.py` to `_prims_common`. Correctness is tested with OpInfo tests (see `PythonRefInfo("ops.nvprims.var_mean"`).

Layer norm reference now uses `torch.var_mean` instead of `torch._refs.var_mean` to allow interception. Here's a simple comparison of performance with this PR and master (on 3080ti):
```py
import torch
from torch._prims.context import TorchRefsNvfuserCapabilityMode
from torch.fx.experimental.proxy_tensor import make_fx
from torch._prims.executor import execute

def func(a):
    return torch.native_layer_norm(a, (1024,), None, None, 1e-6)

a = torch.randn(10, 512, 1024, dtype=torch.float16, device="cuda")

with TorchRefsNvfuserCapabilityMode():
    gm = make_fx(func)(a)

for _ in range(10):
    execute(gm, a, executor="strictly_nvfuser");
```
run with `PYTORCH_NVFUSER_DUMP=dump_eff_bandwidth python script.py`
```py
# WITH THIS PR
# kernel1 run in 0.032768 ms, achieved: 641.25 GB/s
# kernel1 run in 0.033792 ms, achieved: 621.818 GB/s
# kernel1 run in 0.032768 ms, achieved: 641.25 GB/s
# kernel1 run in 0.032608 ms, achieved: 644.396 GB/s
# kernel1 run in 0.031744 ms, achieved: 661.935 GB/s
# kernel1 run in 0.031744 ms, achieved: 661.935 GB/s
# kernel1 run in 0.032768 ms, achieved: 641.25 GB/s
# kernel1 run in 0.03072 ms, achieved: 684 GB/s
# kernel1 run in 0.031744 ms, achieved: 661.935 GB/s
# kernel1 run in 0.031744 ms, achieved: 661.935 GB/s

# ON MASTER
# kernel1 run in 0.05632 ms, achieved: 373.091 GB/s
# kernel1 run in 0.044032 ms, achieved: 477.209 GB/s
# kernel1 run in 0.044032 ms, achieved: 477.209 GB/s
# kernel1 run in 0.044032 ms, achieved: 477.209 GB/s
# kernel1 run in 0.043808 ms, achieved: 479.649 GB/s
# kernel1 run in 0.043008 ms, achieved: 488.571 GB/s
# kernel1 run in 0.044032 ms, achieved: 477.209 GB/s
# kernel1 run in 0.043008 ms, achieved: 488.571 GB/s
# kernel1 run in 0.043008 ms, achieved: 488.571 GB/s
# kernel1 run in 0.043008 ms, achieved: 488.571 GB/s
```
So this PR gives about 35% improvement in performance using nvfuser executor with this specific normalized shape.

Also this PR fixes https://github.com/pytorch/pytorch/issues/83506 (see the change in `torch/csrc/jit/python/pybind_utils.cpp`).

Ref. https://github.com/pytorch/pytorch/issues/80187

@diff-train-skip-merge

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: torch/_prims/context.py [hunk 1]
```diff
@@ -166,15 +167,33 @@ def _is_func_unsupported_nvfuser(torch_function_mode, func, args, kwargs):
...
+        self,
+        orig_func: Callable,
+        types: Sequence,
+        args: Sequence[Any] = (),
+        kwargs: Dict = None,
+    ):
+        if kwargs is None:
+            kwargs = {}
+        # First we intercept calls for nvfuser-specific prims bypassing generic torch._refs
+        if "var_mean" in str(orig_func):
```
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply torch/_prims/context.py:196]: I'll change it to capture only `torch.var_mean` and `torch.ops.aten.var_mean`.

**Comment by IvanYashchuk** [comment]
One failure is real, but it's not reproduced locally.
```
FAIL [0.004s]: test_no_ref_cycle (__main__.FakeTensorConverterTest)
```

**Comment by IvanYashchuk** [comment]
@eellison, could you please take a look at this PR? What's causing this failure?
```
FAIL [0.004s]: test_no_ref_cycle (__main__.FakeTensorConverterTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_fake_tensor.py", line 478, in test_no_ref_cycle
    def test_no_ref_cycle(self):
  File "test_fake_tensor.py", line 478, in test_no_ref_cycle
    def test_no_ref_cycle(self):
  File "test_fake_tensor.py", line 478, in test_no_ref_cycle
    def test_no_ref_cycle(self):
  File "test_fake_tensor.py", line 483, in test_no_ref_cycle
    self.assertEqual(len(mode.fake_tensor_converter.tensor_memo), 1)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 2438, in assertEqual
    msg=(lambda generated_msg: f"{generated_msg} : {msg}") if isinstance(msg, str) and self.longMessage else msg,
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_comparison.py", line 1093, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Scalars are not equal!
``` 
I don't see it on my computer.

**Comment by IvanYashchuk** [comment]
@pytorch/pytorch-dev-infra, @janeyx99 why was this PR reverted? 

I see windows failing, but it seems just a flaky failure:
image

**Comment by weiwangmeta** [comment]
> @pytorch/pytorch-dev-infra, @janeyx99 why was this PR reverted?
> 
> I see windows failing, but it seems just a flaky failure: image

This was landed together with a diff that got reverted, the entire stack from that diff had to be reverted.

**Pull request closed without merging**